File: cli/names.go
package main

// Names for worktree directories: adjective + animal pairs, or Animal
// Crossing character names with doubutsuNames on. The pools are
// embedded from embed/. The app picks no names itself: it asks
// `sm worktrees destination` for one, and `sm create` picks the same way.

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"math/rand/v2"
	"sync"
)

//go:embed embed/name-words.json
var nameWordsJSON []byte

//go:embed embed/doubutsu-names.json
var doubutsuNamesJSON []byte

var nameWords struct {
	Adjectives []string `json:"adjectives"`
	Animals    []string `json:"animals"`
}

func init() {
	if err := json.Unmarshal(nameWordsJSON, &nameWords); err != nil {
		panic("embedded name-words.json is invalid: " + err.Error())
	}
}

// Parsed on first use: only a create without a name needs it.
var doubutsuNames = sync.OnceValue(func() []string {
	var doc struct {
		Names []string `json:"names"`
	}
	if err := json.Unmarshal(doubutsuNamesJSON, &doc); err != nil {
		panic("embedded doubutsu-names.json is invalid: " + err.Error())
	}
	return doc.Names
})

func doubutsuNamesEnabled(global globalConfig) bool {
	return global.DoubutsuNames != nil && *global.DoubutsuNames
}

func namePool(doubutsu bool) []string {
	if doubutsu {
		return doubutsuNames()
	}
	pairs := make([]string, 0, len(nameWords.Adjectives)*len(nameWords.Animals))
	for _, adj := range nameWords.Adjectives {
		for _, animal := range nameWords.Animals {
			pairs = append(pairs, adj+"-"+animal)
		}
	}
	return pairs
}

func pickWorktreeName(used map[string]bool, doubutsu bool) string {
	pool := namePool(doubutsu)
	var candidates []string
	for _, name := range pool {
		if !used[name] {
			candidates = append(candidates, name)
		}
	}
	if len(candidates) > 0 {
		return candidates[rand.IntN(len(candidates))]
	}
	base := pool[rand.IntN(len(pool))]
	for i := 2; ; i++ {
		candidate := fmt.Sprintf("%s-%d", base, i)
		if !used[candidate] {
			return candidate
		}
	}
}

File: host/lib/worktrees/names.ts
// Names for worktree directories. By default an adjective + animal
// pair, like `snuggly-otter` or `zippy-quokka`. With the device's
// `doubutsuNames` setting on, an Animal Crossing villager or character
// name, like `raymond` or `tom-nook`. The branch is what occupies the
// worktree; the name is the "container". Names are single lowercase
// kebab-case tokens so they're safe as filesystem path components.
//
// The pools live in cli/embed/, embedded into the Go CLI (which does
// the picking at create time) and imported here for the New Worktree
// form's pre-pick, so both draw from the same names.
import doubutsuNames from "../../../cli/embed/doubutsu-names.json";
import nameWords from "../../../cli/embed/name-words.json";

const ADJECTIVES: readonly string[] = nameWords.adjectives;
const ANIMALS: readonly string[] = nameWords.animals;
const DOUBUTSU: readonly string[] = doubutsuNames.names;

export function pickWorktreeName(
  used: ReadonlySet<string>,
  doubutsu: boolean,
): string {
  // Built per pick rather than held for the app's lifetime: the form's
  // pre-pick is rare, and enumerating the thousands of pairs lets us
  // pick uniformly across the unused set without retry loops.
  const pool = doubutsu
    ? DOUBUTSU
    : ADJECTIVES.flatMap((adj) => ANIMALS.map((animal) => `${adj}-${animal}`));
  const candidates = pool.filter((name) => !used.has(name));
  if (candidates.length > 0) {
    return candidates[Math.floor(Math.random() * candidates.length)];
  }
  // Every name is taken (vanishingly unlikely for pairs, a few hundred
  // live worktrees for doubutsu names): pick any base and walk numeric
  // suffixes until one's free.
  const base = pool[Math.floor(Math.random() * pool.length)];
  for (let i = 2; ; i++) {
    const candidate = `${base}-${i}`;
    if (!used.has(candidate)) return candidate;
  }
}
